### PR TITLE
CRI: Stop following container log when container exited.

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes https://github.com/kubernetes/kubernetes/issues/44340.

This PR changed kubelet to periodically check whether container is running when following container logs, and stop following when container exited.

I've tried this PR in my local cluster:
```
Wed Apr 12 20:23:54 UTC 2017
Wed Apr 12 20:23:58 UTC 2017
Wed Apr 12 20:24:02 UTC 2017
Wed Apr 12 20:24:06 UTC 2017
Wed Apr 12 20:24:10 UTC 2017
Wed Apr 12 20:24:14 UTC 2017
Wed Apr 12 20:24:18 UTC 2017
Wed Apr 12 20:24:22 UTC 2017
Wed Apr 12 20:24:26 UTC 2017
Wed Apr 12 20:24:30 UTC 2017
Wed Apr 12 20:24:34 UTC 2017
Wed Apr 12 20:24:38 UTC 2017
Wed Apr 12 20:24:42 UTC 2017
Wed Apr 12 20:24:46 UTC 2017
failed to wait logs for log file "/var/log/pods/1d54634c7b31346fc3219f5e0b7507cc/nginx_0.log": container "b9a17a2c53550c3703ab350d85911743af8bf164a41813544fd08fb9585f7501" is not running (state="CONTAINER_EXITED")
```

The only difference is that `ReadLogs` will return error when container exits during following. I'm not sure whether we should get rid of it or not.

@yujuhong @feiskyer @JorritSalverda
/cc @kubernetes/sig-node-bugs 

**Release note**:
```release-note
CRI: `kubectl logs -f` now stops following when container stops, as it did pre-CRI.
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 1]
```diff
@@ -196,6 +204,40 @@ func ReadLogs(path string, apiOpts *v1.PodLogOptions, stdout, stderr io.Writer)
...
+			if errRetry == 0 {
+				return err
+			}
+			errRetry--
+		case <-time.After(stateCheckPeriod):
+			s, err := m.runtimeService.ContainerStatus(id)
+			if err != nil {
+				return err
+			}
+			// Only keep following container log when it is running. There may be a race condition
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: If the container is not running, we can't tail it anyway. Do we need to care about that?

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 2]
```diff
@@ -196,6 +204,40 @@ func ReadLogs(path string, apiOpts *v1.PodLogOptions, stdout, stderr io.Writer)
...
+			// that the container is created but not running yet, since we've already waited for
+			// stateCheckPeriod here, it should be rare.
+			// Even when that happen, user could also retry after the container becomes running.
+			if s.State != runtimeapi.ContainerState_CONTAINER_RUNNING {
+				return fmt.Errorf("container %q is not running (state=%q)", id, s.State)
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Don't think we need the error message.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 3]
```diff
@@ -196,6 +204,40 @@ func ReadLogs(path string, apiOpts *v1.PodLogOptions, stdout, stderr io.Writer)
 	}
 }
 
+// waitLogs wait for the next log write. If the container is not running
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Is this sentence complete? `If the container is not running`

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: I thought we'd only reach this function if the container is running...

Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: ACK. I tried with docker. `docker logs -f` will also return immediately if the container is in created state.

If so, I think we could also return here, we just shouldn't throw out the error.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Then I need to introduce another error to tell `waitLogs` not to continue parse log.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Sorry, must be distracted by something else at that time...

**Comment by mrunalp** [comment]
Can't we continue to tail the logs even if container exited? The log file should be just there, right?

**Comment by Random-Liu** [comment]
@mrunalp We can still tail the log of exited container. However, we should not keep waiting for new lines in following mode, because the container has exited already.

**Comment by JorritSalverda** [comment]
The behaviour prior to 1.6.0 was to stop following logs as soon as the container exited. With an exit code of 0, at least when a container terminated successfully. I used it to tail containers started by a job until successful completion.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Done

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Done

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Done

**Comment by Random-Liu** [comment]
@yujuhong Updated the PR and get rid of the error.
Here is the new output:
```
$ cluster/kubectl.sh logs -f nginx-3
Thu Apr 13 18:23:42 UTC 2017
Thu Apr 13 18:23:46 UTC 2017
Thu Apr 13 18:23:50 UTC 2017
Thu Apr 13 18:23:54 UTC 2017
Thu Apr 13 18:23:58 UTC 2017
Thu Apr 13 18:24:02 UTC 2017
Thu Apr 13 18:24:06 UTC 2017
Thu Apr 13 18:24:10 UTC 2017
$
```

**Comment by yujuhong** [comment]
/lgtm

@Random-Liu could you add a release-note so that we can patch 1.6 later.

**Comment by Random-Liu** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**